src/main/scala/ch/epfl/bluebrain/nexus/kg/search/QueryResultEncoder.scala

```diff
@@ -13,39 +20,74 @@ object QueryResultEncoder {
     * Encoder for scored query results
     */
   implicit val scoredEncoder: Encoder[ScoredQueryResults[AbsoluteIri]] = Encoder.instance { results =>
-    Json
-      .obj(
-        "total"    -> Json.fromLong(results.total),
-        "maxScore" -> Json.fromFloatOrString(results.maxScore),
-        "results" -> Json.arr(
-          results.results.map { res =>
-            Json.obj(
-              "resultId" -> Json.fromString(res.source.show)
-            )
-
-          }: _*
-        )
-      )
-      .addContext(Contexts.searchCtxUri)
+    //not used, but without it Jena will embed everything in @graph
+    val mainNode = url"http://localhost"
+    val graph = Graph(
+      (mainNode, nxv.total, Node.literal(results.total)),
+      (mainNode, nxv.maxScore, Node.literal(results.maxScore))
+    )
+
+    addResults(graph, mainNode, results, queryResultToTriples)
+
   }
 
   /**
     * Encoder for unscored query results
     */
   implicit val unscoredEncoder: Encoder[UnscoredQueryResults[AbsoluteIri]] = Encoder.instance { results =>
-    Json
-      .obj(
-        "total" -> Json.fromLong(results.total),
-        "results" -> Json.arr(
-          results.results.map { res =>
-            Json.obj(
-              "resultId" -> Json.fromString(res.source.show)
-            )
-
-          }: _*
-        )
+    //not used, but without it Jena will embed everything in @graph
+    val mainNode = url"http://localhost"
+    val graph = Graph(
+      (mainNode, nxv.total, Node.literal(results.total))
+    )
+    addResults(graph, mainNode, results, queryResultToTriples)
+  }
+
+  private def queryResultToTriples(result: QueryResult[AbsoluteIri], node: BNode): Set[Graph.Triple] = result match {
+    case ScoredQueryResult(score, source) =>
+      Set(
+        (node, nxv.resultId, Node.literal(source.show)),
+        (node, nxv.score, Node.literal(score))
+      )
+    case UnscoredQueryResult(source) =>
+      Set(
+        (node, nxv.resultId, Node.literal(source.show))
       )
-      .addContext(Contexts.searchCtxUri)
   }
 
+  private def addResults[A <: QueryResult[AbsoluteIri]](
+      graph: Graph,
+      mainNode: IriOrBNode,
+      results: QueryResults[AbsoluteIri],
+      triples: (QueryResult[AbsoluteIri], BNode) => Set[Graph.Triple]): Json = {
+    val graphWithList = results.results match {
+      case Nil => graph.add(mainNode, nxv.results, rdf.nil)
+      case first :: rest =>
+        val firstNode     = BNode()
```

Comment:
this case block is the same as the case block inside the fold left. Why don't you provide an inner function and refactor it together?

Reply:
It's not really the same, here it links the main node with the first list node and inside `foldLeft` it links two list nodes together.